nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearch.java

```diff
@@ -0,0 +1,231 @@
+            .name("Type")
+            .description("The type of this document (used by Elasticsearch for indexing and searching)")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(
+                    AttributeExpression.ResultType.STRING, true))
+            .build();
+
+    public static final PropertyDescriptor BATCH_SIZE = new PropertyDescriptor.Builder()
+            .name("Batch Size")
+            .description("The preferred number of FlowFiles to put to the database in a single transaction")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("100")
+            .build();
+
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        relationships.add(REL_RETRY);
+        return Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(CLUSTER_NAME);
+        descriptors.add(HOSTS);
+        descriptors.add(PROP_SSL_CONTEXT_SERVICE);
+        descriptors.add(PROP_SHIELD_LOCATION);
+        descriptors.add(PING_TIMEOUT);
+        descriptors.add(SAMPLER_INTERVAL);
+        descriptors.add(ID_ATTRIBUTE);
+        descriptors.add(INDEX);
+        descriptors.add(TYPE);
+        descriptors.add(BATCH_SIZE);
+
+        return Collections.unmodifiableList(descriptors);
+    }
+
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        // Create the client if one does not already exist
+        createClient(context);
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions().getValue();
+        final String id_attribute = context.getProperty(ID_ATTRIBUTE).getValue();
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions().getValue();
+
+        final List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        final ProcessorLog logger = getLogger();
+
+        try {
+            final BulkRequestBuilder bulk = esClient.get().prepareBulk();
+            for (FlowFile file : flowFiles) {
+                final String id = file.getAttribute(id_attribute);
+                if (id == null) {
+                    getLogger().error("no value in identifier attribute {}", new Object[]{id_attribute});
+                    throw new ProcessException("No value in identifier attribute " + id_attribute);
```

Comment:
Rather than throwing a ProcessException, it probably makes more sense to just route this FlowFile to failure. This way, if a single FlowFile is missing the necessary attribute, we don't have to roll back the entire ProcessSession. As-is, if one FlowFile is missing the attribute, it can prevent any FlowFile from getting sent because the session will continually get rolled back.